site/docs/v1/tech/migration-guide/firebase.adoc

```diff
@@ -0,0 +1,390 @@
+* <<What to Do Next>>
+* <<Additional Support>>
+
+There are a number of different ways applications can be integrated with
+Firebase, and it would be difficult to cover them all. This guide
+mentions the typical parts of a bulk migration and focuses on migrating
+user data from a Firebase managed user database into FusionAuth.
+
+== Planning Considerations
+
+=== Obtaining User Data
+
+You can use the Firebase CLI or API to export user data. This guide uses
+the CLI, but if you have a large number of users, or want to build the 
+migration into your application, the API might be useful. 
+Firebase also has a https://firebase.google.com/docs/firestore/manage-data/export-import[high level export guide worth
+reviewing].
+
+If you haven't already installed and set up the Firebase CLI, you can
+follow the
+https://firebase.google.com/docs/cli#install-cli-mac-linux[guide here]
+to do so.
+
+
+=== Mapping User Attributes
+
+include::docs/v1/tech/migration-guide/_mapping-user-attributes.adoc[]
+
+=== Social Logins
+
+include::docs/v1/tech/migration-guide/_social-login-note.adoc[]
+
+include::docs/v1/tech/migration-guide/_social-login-migration.adoc[]
+
+=== Other Entities
+
+:other_migrated_entities: connections or roles
+include::docs/v1/tech/migration-guide/_other-entities-intro.adoc[]
+:other_migrated_entities!:
+
+* In Firebase, sign-in providers are a source of data for users.
+FusionAuth calls these Identity Providers.
+* Observers are ways for you to customize authentication
+or authorization workflows. FusionAuth has a similar concept called
+Lambdas.
+* With Firebase, Applications are what your
+users can log in to. FusionAuth refers to these as Applications as well.
+* Projects are a high level construct which groups entities such as
+users and applications together. FusionAuth calls these Tenants.
+* For Firebase, Roles can be implemented using a combination of https://firebase.google.com/docs/auth/admin/custom-claims[custom claims and security rules.] FusionAuth has Roles and they are defined on an Application by Application basis.
+* Refresh tokens allow JWTs to be refreshed without a user logging in.
+These can be migrated using the link:/docs/v1/tech/apis/users#import-refresh-tokens[Import Refresh Tokens API].
+* Firebase has many other products your applications might be using, from Cloud Firestore for data, Analytics and more. FusionAuth does not have equivalents for these products, and is focused only on authentication. You can use FusionAuth in conjunction with all the other Firebase products.
+
+[NOTE]
+====
+In FusionAuth, users are explicitly mapped to applications with a link:/docs/v1/tech/core-concepts/registrations[Registration].
+
+In Firebase, users have access to all Firebase Applications in the same project by default. You can control access by using Roles and Security Rules per Application if you need.
+====
+
+
+==== Identifiers 
+
+include::docs/v1/tech/migration-guide/_identifiers.adoc[]
```

Comment:
they moved to UUIDs in 2016, but now away again, so we tweaked this a bit. This would be a major pain for people migrating from Firebase if they need to update all external dependencies to link to the new FusionAuth UUID? Or is there a workaround we should mention here.

Reply:
The typical workaround is:

* let FusionAuth own the Ids
* add a `user.data.originalId` field, with the value of the firebase id.

Everything under `user.data` is indexed and available via search and so the system can look up a user by either their FusionAuth id or the firebase Id.

Would be great to mention this workaround. You could add it below the included generic file, if it reads okay.